tests/main/hidden-snap-dir/task.yaml

```diff
@@ -0,0 +1,83 @@
+summary: Check that the experimental hidden dir feature migrates the dir
+
+environment:
+    NAME: test-snapd-tools
+
+prepare: |
+    snap pack "$TESTSLIB"/snaps/"$NAME"
+    "$TESTSTOOLS"/snaps-state install-local "$NAME"
+
+restore: |
+    snap set system experimental.hidden-snap-folder!
+
+execute: |
+    echo "Set experimental hidden snap folder feature"
+    snap set system experimental.hidden-snap-folder=true
+
+    echo "Check that nothing has been migrated yet"
+    snapEnv=$("$NAME".env)
+    echo "$snapEnv" | MATCH "SNAP_USER_DATA=/root/snap/$NAME/x1"
+    echo "$snapEnv" | MATCH "SNAP_USER_COMMON=/root/snap/$NAME/common"
+
+    test -d "$HOME"/snap
+    not test -d "$HOME"/.snap/data
+
+    echo "Take a snapshot"
+    "$NAME".cmd echo "prev_data" > "$HOME"/snap/"$NAME"/current/data
+    snapshot=$(snap save "$NAME" | awk 'FNR == 2 {print $1}')
+
+    echo "Write data to user data dirs"
+    "$NAME".cmd echo "data" > "$HOME"/snap/"$NAME"/current/data
+    "$NAME".cmd echo "common" > "$HOME"/snap/"$NAME"/common/common
+
+    echo "Refresh the snap"
+    "$TESTSTOOLS"/snaps-state install-local "$NAME"
+
+    echo "Check snap directory was migrated"
+    test -d "$HOME"/.snap/data/"$NAME"
+    test -d "$HOME"/.snap/data/"$NAME"/common
+    test -d "$HOME"/.snap/data/"$NAME"/x2
+    not test -d "$HOME"/snap/"$NAME"
+
+    echo "Check the env vars point to ~/.snap/data"
+    snapEnv=$("$NAME".env)
+    echo "$snapEnv" | MATCH "SNAP_USER_DATA=$HOME/\.snap/data/$NAME/x2"
+    echo "$snapEnv" | MATCH "SNAP_USER_COMMON=$HOME/\.snap/data/$NAME/common"
+
+    # 'current' symlink is created just before the snap runs, so this check
+    # must come after a snap run
+    if [ "$(readlink "$HOME"/.snap/data/"$NAME"/current)" != "x2" ]; then
+      echo "expected 'current' to point to new revision after refresh"
+      exit 1
+    fi
+
+    echo "Check that the written data is in the new dir"
+    # shellcheck disable=SC2002
+    cat "$HOME"/.snap/data/"$NAME"/common/common | MATCH "common"
+    # shellcheck disable=SC2002
+    cat "$HOME"/.snap/data/"$NAME"/x2/data | MATCH "data"
+
+    echo "Check the snap can write to the new dirs"
+    #shellcheck disable=SC2016
+    "$NAME".cmd sh -c 'echo "new_data" > "$SNAP_USER_DATA"/new_data'
+    #shellcheck disable=SC2016
+    "$NAME".cmd sh -c 'echo "new_common" > "$SNAP_USER_COMMON"/new_common'
+    # shellcheck disable=SC2002
+    cat "$HOME"/.snap/data/"$NAME"/common/new_common | MATCH "new_common"
+    # shellcheck disable=SC2002
+    cat "$HOME"/.snap/data/"$NAME"/x2/new_data | MATCH "new_data"
+
+    echo "Restore snapshot and check data was restored"
+    snap restore "$snapshot"
+    # shellcheck disable=SC2002
+    cat "$HOME"/.snap/data/"$NAME"/x2/data | MATCH "prev_data"
```

Comment:
(nitpick) `MATCH "prev_data" < "$HOME"/.snap/data/"$NAME"/x2/data ` would also work and avoid the shellcheck disable - or am I missing something? Not really important just a tiny bit more compact because it avoids the extra `shellcheck disable` line.

Reply:
No, that's good too. I just went with this option because in a previous PR someone preferred the cat and UUoC warning disable since it was more explicit. But this one is very simple, so I think `<` is quite readable.